p2p/pex/pex_reactor.go

```diff
@@ -395,13 +395,35 @@ func (r *Reactor) ReceiveAddrs(addrs []*p2p.NetAddress, src Peer) error {
 
 // SendAddrs sends addrs to the peer.
 func (r *Reactor) SendAddrs(p Peer, netAddrs []*p2p.NetAddress) {
+	msg := filterAddrs(netAddrs)
 	e := p2p.Envelope{
 		ChannelID: PexChannel,
-		Message:   &tmp2p.PexAddrs{Addrs: p2p.NetAddressesToProto(netAddrs)},
+		Message:   msg,
 	}
 	p.Send(e)
 }
 
+// safeMaxAddressEstimationMargin is a factor used to estimate the maximum number of addresses fitting into a size limit.
+const safeMaxAddressEstimationMargin = 0.9
+
+// filterAddrs transforms a slice of NetAddress into PexAddrs while ensuring the resulting message size doesn't exceed maxMsgSize.
+func filterAddrs(netAddrs []*p2p.NetAddress) *tmp2p.PexAddrs {
+	msg := tmp2p.PexAddrs{Addrs: p2p.NetAddressesToProto(netAddrs)}
+	if msg.Size() > maxMsgSize {
+		// Calculate how many addresses we can fit based on the ratio
+		avgAddrSize := msg.Size() / len(netAddrs)
+		estimatedMaxAddrs := (maxMsgSize * safeMaxAddressEstimationMargin) / avgAddrSize
+
+		if estimatedMaxAddrs <= 0 {
+			estimatedMaxAddrs = 1
+		}
+
+		msg = tmp2p.PexAddrs{Addrs: p2p.NetAddressesToProto(netAddrs[:estimatedMaxAddrs])}
+	}
+
+	return &msg
+}
```

Comment:
Couldn't you just for loop removing addresses until it fit the limit rather than having to have a magic number? @rach-id 

And this method would be better called `capAddresses` or `limitAddresses`. Filter doesn't tell us what it's filtering for

Reply:
> Couldn't you just for loop removing addresses until it fit the limit rather than having to have a magic number?

Since this will be called multiple times, I decided to go for this approach as it's faster. I can switch if you prefer